#### Loading all the required libraries 

In [1]:
import pandas as pd
import glob
import os
from ast import literal_eval
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

from sqlalchemy import create_engine


#### Automating some of the things by loading all the path for the files and creating df names from them followed by loading the actually data to the respective dataframes.

In [2]:
filepaths = [i for i in os.listdir(".") if i.endswith('.gz')]
filepaths

['brands.json.gz', 'receipts.json.gz', 'users.json.gz']

In [3]:
df_to_create = [i.replace('.json.gz','_df') for i in filepaths]
df_to_create

['brands_df', 'receipts_df', 'users_df']

In [4]:
for i,j in zip(df_to_create,filepaths):
    exec('{0} = pd.read_json("{1}",compression="infer",lines=True)'.format(i,j))
     

# Brand_DF

In [5]:
brands_df.head(50)

_id       barcode              category  \
0   {'$oid': '601ac115be37ce2ead437551'}  511111019862                Baking   
1   {'$oid': '601c5460be37ce2ead43755f'}  511111519928             Beverages   
2   {'$oid': '601ac142be37ce2ead43755d'}  511111819905                Baking   
3   {'$oid': '601ac142be37ce2ead43755a'}  511111519874                Baking   
4   {'$oid': '601ac142be37ce2ead43755e'}  511111319917        Candy & Sweets   
5   {'$oid': '601ac142be37ce2ead43755b'}  511111719885                Baking   
6   {'$oid': '601ac142be37ce2ead43755c'}  511111219897                Baking   
7   {'$oid': '5cdad0f5166eb33eb7ce0faa'}  511111104810   Condiments & Sauces   
8   {'$oid': '5ab15636e4b0be0a89bb0b07'}  511111504412  Canned Goods & Soups   
9   {'$oid': '5c408e8bcd244a1fdb47aee7'}  511111504788                Baking   
10  {'$oid': '5f4bf556be37ce0b4491554d'}  511111516354                Baking   
11  {'$oid': '57c08106e4b0718ff5fcb02c'}  511111102540                   NaN   
12  {'$oid': '588ba07be4b02187f85cdadd'}  511111201076                Baking   
13  {'$oid': '5d6413156d5f3b23d1bc790a'}  511111205012             Magazines   
14  {'$oid': '585a9611e4b03e62d1ce0e74'}  511111801801    Breakfast & Cereal   
15  {'$oid': '57e5820ce4b0ac389136a311'}  511111202233     Beer Wine Spirits   
16  {'$oid': '5fb6adb8be37ce522e165cb8'}  511111817376     Health & Wellness   
17  {'$oid': '5f358338be37ce443bf9d55a'}  511111515319                Baking   
18  {'$oid': '5fb28549be37ce522e165cb5'}  511111317364                Baking   
19  {'$oid': '592486bfe410d61fcea3d139'}  511111300700                Beauty   
20  {'$oid': '5c4699f387ff3577e203ea29'}  511111305125                  Baby   
21  {'$oid': '5da6071ea60b87376833e34d'}  511111005650     Health & Wellness   
22  {'$oid': '57ebc011e4b0ac389136a335'}  511111802129   Condiments & Sauces   
23  {'$oid': '5332f5fee4b03c9a25efd0bd'}  511111303947                   NaN   
24  {'$oid': '5332fa7ce4b03c9a25efd22e'}  511111802914                   NaN   
25  {'$oid': '5e9f18bfbe37ce3e45b6a77f'}  511111914549                Baking   
26  {'$oid': '592486bee410d61fcea3d133'}  511111400769                Frozen   
27  {'$oid': '5d66d71fa3a018093ab34728'}  511111105329             Magazines   
28  {'$oid': '5f493e72be37ce64d0ae36c6'}  511111316114                Baking   
29  {'$oid': '5f4936ddbe37ce52f8314fd9'}  511111315957                Baking   
30  {'$oid': '57ebc2e7e4b0ac389136a34b'}  511111201915               Grocery   
31  {'$oid': '5fd2a0aebe37ce49eb72c0ee'}  511111518112                Baking   
32  {'$oid': '5dc03596a60b873d6b0666cd'}  511111506249    Breakfast & Cereal   
33  {'$oid': '5f494c5f04db711dd8fe87e6'}  511111916161                Baking   
34  {'$oid': '5332f772e4b03c9a25efd125'}  511111103653                   NaN   
35  {'$oid': '59dfaad1e4b0a56a2fa69abc'}  511111100621                Baking   
36  {'$oid': '5daf33e71dda2c3e1416ae93'}  511111305910                Snacks   
37  {'$oid': '5f35a0babe37ce6853cff1ff'}  511111115366                Baking   
38  {'$oid': '5fd29590be37ce38bbbefb28'}  511111718079                Baking   
39  {'$oid': '5fd3acb6be37ce0d9386bb40'}  511111818182                Baking   
40  {'$oid': '5bd200a6965c7d66d92731ea'}  511111504627             Household   
41  {'$oid': '585a9675e4b03e62d1ce0e7f'}  511111301691             Beverages   
42  {'$oid': '5887a17ee4b02187f85cdad3'}  511111101178             Beverages   
43  {'$oid': '5332f765e4b03c9a25efd11f'}  511111503699                   NaN   
44  {'$oid': '5d66d94d6d5f3b6188d4f04b'}  511111505365             Magazines   
45  {'$oid': '5d66d2fd6d5f3b6188d4f045'}  511111405252             Magazines   
46  {'$oid': '5332f5f3e4b03c9a25efd0ad'}  511111104025                   NaN   
47  {'$oid': '5fa1c567be37ce402c4618f1'}  511111317203                Baking   
48  {'$oid': '5a5d1e76e4b0db471c2d0425'}  511111900337     Beer Wine Spirits   
49  {'$oid': '55b630d9e4b0d8e685c1421

In [6]:
brands_df.shape

(1167, 8)

In [7]:
brands_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           1167 non-null   object 
 1   barcode       1167 non-null   int64  
 2   category      1012 non-null   object 
 3   categoryCode  517 non-null    object 
 4   cpg           1167 non-null   object 
 5   name          1167 non-null   object 
 6   topBrand      555 non-null    float64
 7   brandCode     933 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 73.1+ KB


Creating a dummy dataframe to clean all the missing information or imputing some new information, as working with the original dataframe can cause an issue.

In [8]:
brands_df_copy = brands_df.copy()

Created this method to find all the keys available inside a dictionary

In [9]:
def findKeys(df,col):
    if(df[col].isnull().sum() == 0):
        return set().union(*(d.keys() for d in df[col]))
    else:
        return set().union(*(d.keys() for d in df.dropna(subset=[col])[col]))

As for an instance we can see, we have following two key within our dataframe, id and ref.

In [10]:
brands_df['cpg'][0]

{'$id': {'$oid': '601ac114be37ce2ead437550'}, '$ref': 'Cogs'}

In [11]:
brands_df_copy['_id'] = brands_df_copy['_id'].apply(lambda x: x['$oid'])

In [12]:
findKeys(brands_df_copy,'cpg')

{'$id', '$ref'}

Using **json_normalize** method from pandas to get the semi-structured JSON data into a flat table 

In [13]:
brands_cpg_norm = pd.json_normalize(brands_df_copy['cpg']).add_prefix('cpg_')
brands_cpg_norm

cpg_$ref              cpg_$id.$oid
0        Cogs  601ac114be37ce2ead437550
1        Cogs  5332f5fbe4b03c9a25efd0ba
2        Cogs  601ac142be37ce2ead437559
3        Cogs  601ac142be37ce2ead437559
4        Cogs  5332fa12e4b03c9a25efd1e7
...       ...                       ...
1162     Cogs  5f77274dbe37ce6b592e90bf
1163     Cogs  53e10d6368abd3c7065097cc
1164     Cogs  5332fa12e4b03c9a25efd1e7
1165     Cogs  5332f5f6e4b03c9a25efd0b4
1166     Cogs  6026d757be37ce6369301467

[1167 rows x 2 columns]

As we have already got all the data from **cpg** column, so we are dropping it.

In [14]:
brands_df_copy.drop(columns='cpg',inplace=True)

#### Finally merging both the dataframe and creating a final clean dataframe.

In [15]:
brands_final = pd.merge(brands_df_copy, brands_cpg_norm, left_index=True, right_index=True, how='outer')
brands_final.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                        name  topBrand                      brandCode  \
0  test brand @1612366101024       0.0                            NaN   
1                  Starbucks       0.0                      STARBUCKS   
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176   
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051   
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827   

  cpg_$ref              cpg_$id.$oid  
0     Cogs  601ac114be37ce2ead437550  
1     Cogs  5332f5fbe4b03c9a25efd0ba  
2     Cogs  601ac142be37ce2ead437559  
3     Cogs  601ac142be37ce2ead437559  
4     Cogs  5332fa12e4b03c9a25efd1e7

# Receipt_DF

In [16]:
receipts_df.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [17]:
receipts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      1119 non-null   object 
 1   bonusPointsEarned        544 non-null    float64
 2   bonusPointsEarnedReason  544 non-null    object 
 3   createDate               1119 non-null   object 
 4   dateScanned              1119 non-null   object 
 5   finishedDate             568 non-null    object 
 6   modifyDate               1119 non-null   object 
 7   pointsAwardedDate        537 non-null    object 
 8   pointsEarned             609 non-null    float64
 9   purchaseDate             671 non-null    object 
 10  purchasedItemCount       635 non-null    float64
 11  rewardsReceiptItemList   679 non-null    object 
 12  rewardsReceiptStatus     1119 non-null   object 
 13  totalSpent               684 non-null    float64
 14  userId                  

In [18]:
receipts_df.shape

(1119, 15)

Following the same methodology of creating a dummy DF and cleaning it

In [19]:
receipts_df_copy = receipts_df.copy()

In [20]:
receipts_df_copy['_id'] = receipts_df_copy['_id'].apply(lambda x: x['$oid'])

Creating a function to convert all the **date columns** from **UTC to datetime format**.

In [21]:
def date_conv(x):
    try:
        return(datetime.utcfromtimestamp(int(x['$date'])/1000).strftime('%Y-%m-%d %H:%M:%S'))
    except TypeError:
        return(None)

In [22]:
receipts_df_copy['createDate'] = receipts_df_copy['createDate'].apply(lambda x: date_conv(x))
receipts_df_copy['dateScanned'] = receipts_df_copy['dateScanned'].apply(lambda x: date_conv(x))
receipts_df_copy['finishedDate'] = receipts_df_copy['finishedDate'].apply(lambda x: date_conv(x))
receipts_df_copy['modifyDate'] = receipts_df_copy['modifyDate'].apply(lambda x: date_conv(x))
receipts_df_copy['pointsAwardedDate'] = receipts_df_copy['pointsAwardedDate'].apply(lambda x: date_conv(x))
receipts_df_copy['purchaseDate'] = receipts_df_copy['purchaseDate'].apply(lambda x: date_conv(x))

Now most of the receipt dataframe columns are clean, only column left is *rewardReceiptItemList*. Working with this columns was complex as it has multiple field which itself could be a column.

But there were many records which comprised of two records within the list. So we have to expand this list such that we have only one record per row.

For example, **receipts_df_copy['rewardsReceiptItemList'][1]** this row has two records as we can see two entries for barcode variable.


In [23]:
receipts_df_copy.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason           createDate  \
0  Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2                         All-receipts receipt bonus  2021-01-03 15:25:37   
3                         All-receipts receipt bonus  2021-01-03 15:25:34   
4                         All-receipts receipt bonus  2021-01-03 15:25:06   

           dateScanned         finishedDate           modifyDate  \
0  2021-01-03 15:25:31  2021-01-03 15:25:31  2021-01-03 15:25:36   
1  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
2  2021-01-03 15:25:37                 None  2021-01-03 15:25:42   
3  2021-01-03 15:25:34  2021-01-03 15:25:34  2021-01-03 15:25:39   
4  2021-01-03 15:25:06  2021-01-03 15:25:11  2021-01-03 15:25:11   

     pointsAwardedDate  pointsEarned         purchaseDate  purchasedItemCount  \
0  2021-01-03 15:25:31         500.0  2021-01-03 00:00:00                 5.0   
1  2021-01-03 15:24:43         150.0  2021-01-02 15:24:43                 2.0   
2                 None           5.0  2021-01-03 00:00:00                 1.0   
3  2021-01-03 15:25:34           5.0  2021-01-03 00:00:00                 4.0   
4  2021-01-03 15:25:06           5.0  2021-01-02 15:25:06                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

   totalSpent                    userId  
0        26.0  5ff1e1eacfcf6c399c274ae6  
1        11.0  5ff1e194b6a9d73a3a9f1052  
2        10.0  5ff1e1f1cfcf6c399c274b0b  
3        28.0  5ff1e1eacfcf6c399c274ae6  
4         1.0  5ff1e194b6a9d73a3a9f1052

In [24]:
receipts_df_copy['rewardsReceiptItemList'][0]

[{'barcode': '4011',
  'description': 'ITEM NOT FOUND',
  'finalPrice': '26.00',
  'itemPrice': '26.00',
  'needsFetchReview': False,
  'partnerItemId': '1',
  'preventTargetGapPoints': True,
  'quantityPurchased': 5,
  'userFlaggedBarcode': '4011',
  'userFlaggedNewItem': True,
  'userFlaggedPrice': '26.00',
  'userFlaggedQuantity': 5}]

We are using **df.explode() function** to transform each element from the list to a row data, also this will create duplicate values for all other column

In [25]:
receipts_df_copy = receipts_df_copy.explode('rewardsReceiptItemList')
receipts_df_copy.reset_index(inplace=True)

In [26]:
receipts_df_copy.shape

(7381, 16)

In [27]:
receipts_df_copy.head()

index                       _id  bonusPointsEarned  \
0      0  5ff1e1eb0a720f0523000575              500.0   
1      1  5ff1e1bb0a720f052300056b              150.0   
2      1  5ff1e1bb0a720f052300056b              150.0   
3      2  5ff1e1f10a720f052300057a                5.0   
4      3  5ff1e1ee0a7214ada100056f                5.0   

                             bonusPointsEarnedReason           createDate  \
0  Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
3                         All-receipts receipt bonus  2021-01-03 15:25:37   
4                         All-receipts receipt bonus  2021-01-03 15:25:34   

           dateScanned         finishedDate           modifyDate  \
0  2021-01-03 15:25:31  2021-01-03 15:25:31  2021-01-03 15:25:36   
1  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
2  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
3  2021-01-03 15:25:37                 None  2021-01-03 15:25:42   
4  2021-01-03 15:25:34  2021-01-03 15:25:34  2021-01-03 15:25:39   

     pointsAwardedDate  pointsEarned         purchaseDate  purchasedItemCount  \
0  2021-01-03 15:25:31         500.0  2021-01-03 00:00:00                 5.0   
1  2021-01-03 15:24:43         150.0  2021-01-02 15:24:43                 2.0   
2  2021-01-03 15:24:43         150.0  2021-01-02 15:24:43                 2.0   
3                 None           5.0  2021-01-03 00:00:00                 1.0   
4  2021-01-03 15:25:34           5.0  2021-01-03 00:00:00                 4.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  {'barcode': '4011', 'description': 'ITEM NOT F...             FINISHED   
1  {'barcode': '4011', 'description': 'ITEM NOT F...             FINISHED   
2  {'barcode': '028400642255', 'description': 'DO...             FINISHED   
3  {'needsFetchReview': False, 'partnerItemId': '...             REJECTED   
4  {'barcode': '4011', 'description': 'ITEM NOT F...             FINISHED   

   totalSpent                    userId  
0        26.0  5ff1e1eacfcf6c399c274ae6  
1        11.0  5ff1e194b6a9d73a3a9f1052  
2        11.0  5ff1e194b6a9d73a3a9f1052  
3        10.0  5ff1e1f1cfcf6c399c274b0b  
4        28.0  5ff1e1eacfcf6c399c274ae6

In [28]:
receipts_df_copy.isnull().sum()

index                         0
_id                           0
bonusPointsEarned          1401
bonusPointsEarnedReason    1401
createDate                    0
dateScanned                   0
finishedDate               1411
modifyDate                    0
pointsAwardedDate          1301
pointsEarned               1128
purchaseDate                458
purchasedItemCount          484
rewardsReceiptItemList      440
rewardsReceiptStatus          0
totalSpent                  435
userId                        0
dtype: int64

We are filling NAN with '{}' value for *rewardsReceiptItemList* column, followed by converting each value to be a string and finally using literal_eval, to find the data type of the content stored in this case it is dict.

In [29]:
receipts_df_copy = receipts_df_copy.fillna({'rewardsReceiptItemList':'{}'})
receipts_df_copy['rewardsReceiptItemList'] = receipts_df_copy['rewardsReceiptItemList'].apply(lambda x:str(x))
receipts_df_copy['rewardsReceiptItemList'] = receipts_df_copy['rewardsReceiptItemList'].apply(literal_eval)

In [30]:
receipts_df_copy['rewardsReceiptItemList'][:20]

0     {'barcode': '4011', 'description': 'ITEM NOT F...
1     {'barcode': '4011', 'description': 'ITEM NOT F...
2     {'barcode': '028400642255', 'description': 'DO...
3     {'needsFetchReview': False, 'partnerItemId': '...
4     {'barcode': '4011', 'description': 'ITEM NOT F...
5     {'barcode': '4011', 'description': 'ITEM NOT F...
6     {'barcode': '1234', 'finalPrice': '2.56', 'ite...
7     {'barcode': '4011', 'description': 'ITEM NOT F...
8     {'brandCode': 'MISSION', 'competitorRewardsGro...
9     {'barcode': '046000832517', 'brandCode': 'BRAN...
10    {'barcode': '4011', 'description': 'ITEM NOT F...
11    {'barcode': '4011', 'description': 'ITEM NOT F...
12    {'barcode': '4011', 'description': 'ITEM NOT F...
13    {'barcode': '013562300631', 'description': 'An...
14    {'barcode': '034100573065', 'description': 'MI...
15    {'barcode': '034100573065', 'description': 'MI...
16    {'barcode': '034100573065', 'description': 'MI...
17    {'barcode': '034100573065', 'description':

Again as we used ** pandas.json_normalize() ** method to convert semi-structured JSON to a flat table.

In [31]:
receipts_Reward_norm = pd.json_normalize(receipts_df_copy['rewardsReceiptItemList'],errors='ignore',record_prefix='rewardsReceiptItemList').add_prefix('rewardsReceiptItemList_')

In [32]:
receipts_Reward_norm

rewardsReceiptItemList_barcode  \
0                              4011   
1                              4011   
2                      028400642255   
3                               NaN   
4                              4011   
...                             ...   
7376                            NaN   
7377                            NaN   
7378                     B076FJ92M4   
7379                     B07BRRLSVC   
7380                            NaN   

                     rewardsReceiptItemList_description  \
0                                        ITEM NOT FOUND   
1                                        ITEM NOT FOUND   
2     DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3                                                   NaN   
4                                        ITEM NOT FOUND   
...                                                 ...   
7376                                                NaN   
7377                                                NaN   
7378  mueller austria hypergrind precision electric ...   
7379  thindust summer face mask - sun protection nec...   
7380                                                NaN   

     rewardsReceiptItemList_finalPrice rewardsReceiptItemList_itemPrice  \
0                                26.00                            26.00   
1                                    1                                1   
2                                10.00                            10.00   
3                                  NaN                              NaN   
4                                28.00                            28.00   
...                                ...                              ...   
7376                               NaN                              NaN   
7377                               NaN                              NaN   
7378                             22.97                            22.97   
7379                             11.99                            11.99   
7380                               NaN                              NaN   

     rewardsReceiptItemList_needsFetchReview  \
0                                      False   
1                                        NaN   
2                                       True   
3                                      False   
4                                      False   
...                                      ...   
7376                                     NaN   
7377                                     NaN   
7378                                     NaN   
7379                                     NaN   
7380                                     NaN   

     rewardsReceiptItemList_partnerItemId  \
0                                       1   
1                                       1   
2                                       2   
3                                       1   
4                                       1   
...                                   ...   
7376                                  NaN   
7377                                  NaN   
7378                                    0   
7379                                    1   
7380                                  NaN   

     rewardsReceiptItemList_preventTargetGapPoints  \
0                                             True   
1                                              NaN   
2                                             True   
3                                             True   
4                                             True   
...                                            ...   
7376                                           NaN   
7377                                           NaN   
7378                                           NaN   
7379                                           NaN   
7380                                           NaN   

      rewardsReceiptItemList_quantityPurchased  \
0                                          5.0   
1                                          1.0   
2        

In [33]:
receipts_Reward_norm['rewardsReceiptItemList_barcode'].value_counts()

4011            177
036000320893     92
034100573065     90
036000391718     87
012000809941     76
               ... 
4562              1
021000701445      1
021000026999      1
686924560242      1
665290001184      1
Name: rewardsReceiptItemList_barcode, Length: 568, dtype: int64

In [34]:
receipts_df_copy.drop(columns=['index','rewardsReceiptItemList'],inplace=True)

Merging dummy and normalized dataframe and creating a final clean dataframe.

In [35]:
receipts_final = pd.merge(receipts_df_copy, receipts_Reward_norm, left_index=True, right_index=True, how='outer')
receipts_final.shape

(7381, 48)

In [36]:
receipts_final.head(40)

_id  bonusPointsEarned  \
0   5ff1e1eb0a720f0523000575              500.0   
1   5ff1e1bb0a720f052300056b              150.0   
2   5ff1e1bb0a720f052300056b              150.0   
3   5ff1e1f10a720f052300057a                5.0   
4   5ff1e1ee0a7214ada100056f                5.0   
5   5ff1e1d20a7214ada1000561                5.0   
6   5ff1e1d20a7214ada1000561                5.0   
7   5ff1e1e40a7214ada1000566              750.0   
8   5ff1e1cd0a720f052300056f                5.0   
9   5ff1e1a40a720f0523000569              500.0   
10  5ff1e1ed0a7214ada100056e                5.0   
11  5ff1e1eb0a7214ada100056b              250.0   
12  5ff1e1c50a720f052300056c              100.0   
13  5ff1e1a10a720f0523000568              750.0   
14  5ff1e1b60a7214ada100055c              150.0   
15  5ff1e1b60a7214ada100055c              150.0   
16  5ff1e1b60a7214ada100055c              150.0   
17  5ff1e1b60a7214ada100055c              150.0   
18  5ff1e1b60a7214ada100055c              150.0   
19  5ff1e1b60a7214ada100055c              150.0   
20  5ff1e1b60a7214ada100055c              150.0   
21  5ff1e1b60a7214ada100055c              150.0   
22  5ff1e1b60a7214ada100055c              150.0   
23  5ff1e1b60a7214ada100055c              150.0   
24  5f9c74f70a7214ad07000037              750.0   
25  5f9c74f70a7214ad07000037              750.0   
26  5f9c74f70a7214ad07000037              750.0   
27  5f9c74f70a7214ad07000037              750.0   
28  5f9c74f70a7214ad07000037              750.0   
29  5f9c74f70a7214ad07000037              750.0   
30  5f9c74f70a7214ad07000037              750.0   
31  5f9c74f70a7214ad07000037              750.0   
32  5f9c74f70a7214ad07000037              750.0   
33  5f9c74f70a7214ad07000037              750.0   
34  5f9c74f70a7214ad07000037              750.0   
35  5ff1e1b20a7214ada100055a              300.0   
36  5ff1e1e90a7214ada1000569                NaN   
37  5ff1e1e90a7214ada1000569                NaN   
38  5ff1e1df0a7214ada1000564              750.0   
39  5ff1e1df0a7214ada1000564              750.0   

                              bonusPointsEarnedReason           createDate  \
0   Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1   Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2   Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
3                          All-receipts receipt bonus  2021-01-03 15:25:37   
4                          All-receipts receipt bonus  2021-01-03 15:25:34   
5                          All-receipts receipt bonus  2021-01-03 15:25:06   
6                          All-receipts receipt bonus  2021-01-03 15:25:06   
7   Receipt number 1 completed, bonus point schedu...  2021-01-03 15:25:24   
8                          All-receipts receipt bonus  2021-01-03 15:25:01   
9   Receipt number 2 completed, bonus point schedu...  2021-01-03 15:24:20   
10                         All-receipts receipt bonus  2021-01-03 15:25:33   
11  Receipt number 3 completed, bonus point schedu...  2021-01-03 15:25:31   
12  Receipt number 6 completed, bonus point schedu...  2021-01-03 15:24:53   
13  Receipt number 1 completed, bonus point schedu...  2021-01-03 15:24:17   
14  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
15  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
16  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
17  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
18  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
19  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
20  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
21  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
22  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:38   
23  Receipt number 5 completed, bonus point schedu...  2021-0

In [37]:
receipts_final.columns

Index(['_id', 'bonusPointsEarned', 'bonusPointsEarnedReason', 'createDate',
       'dateScanned', 'finishedDate', 'modifyDate', 'pointsAwardedDate',
       'pointsEarned', 'purchaseDate', 'purchasedItemCount',
       'rewardsReceiptStatus', 'totalSpent', 'userId',
       'rewardsReceiptItemList_barcode', 'rewardsReceiptItemList_description',
       'rewardsReceiptItemList_finalPrice', 'rewardsReceiptItemList_itemPrice',
       'rewardsReceiptItemList_needsFetchReview',
       'rewardsReceiptItemList_partnerItemId',
       'rewardsReceiptItemList_preventTargetGapPoints',
       'rewardsReceiptItemList_quantityPurchased',
       'rewardsReceiptItemList_userFlaggedBarcode',
       'rewardsReceiptItemList_userFlaggedNewItem',
       'rewardsReceiptItemList_userFlaggedPrice',
       'rewardsReceiptItemList_userFlaggedQuantity',
       'rewardsReceiptItemList_needsFetchReviewReason',
       'rewardsReceiptItemList_pointsNotAwardedReason',
       'rewardsReceiptItemList_pointsPayerId',
      

# Users_DF

In [38]:
users_df.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [39]:
users_final = users_df.copy()

Using same data cleaning method to clean this dataframe

In [40]:
users_final['_id'] = users_final['_id'].apply(lambda x: x['$oid'])
users_final['createdDate'] = users_final['createdDate'].apply(lambda x: date_conv(x))
users_final['lastLogin'] = users_final['lastLogin'].apply(lambda x: date_conv(x))

In [41]:
users_final.head()

_id  active          createdDate            lastLogin  \
0  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
1  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
2  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
3  5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30  2021-01-03 15:25:30   
4  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   

       role signUpSource state  
0  consumer        Email    WI  
1  consumer        Email    WI  
2  consumer        Email    WI  
3  consumer        Email    WI  
4  consumer        Email    WI

# First: Review Existing Unstructured Data and Diagram a New Structured Relational Data Model

![Structured Relation Data Model](SQLdiagram.png "Structured Relation Data Model")

# Second: Write a query that directly answers a predetermined question from a business stakeholder

#### Converting our dataframes to SQL database

In [42]:
engine = create_engine('sqlite:///Fetch_rewards.db', echo=False)
sqlite_connection = engine.connect()

In [43]:
brands_final.to_sql('brands', sqlite_connection, if_exists='replace')

In [44]:
receipts_final.to_sql('receipt', sqlite_connection, if_exists='replace')

In [45]:
users_final.to_sql('users', sqlite_connection, if_exists='replace')

In [46]:
sqlite_connection.close()

## What are the top 5 brands by receipts scanned for most recent month?

![](SQL_Answers/Question1Script.png "Question 1")
![](SQL_Answers/Question1Ans.png "Question 1")

# How does the ranking of the top 5 brands by receipts scanned for the recent month compare to the ranking for the previous month?

![](SQL_Answers/Question2Script.png "Question 2")
![](SQL_Answers/Question2Ans.png "Question 2")

## When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [47]:
receipts_final['rewardsReceiptStatus'].value_counts()

FINISHED     5920
FLAGGED       810
SUBMITTED     434
REJECTED      167
PENDING        50
Name: rewardsReceiptStatus, dtype: int64

![](SQL_Answers/Question3Script.png "Question 3")
![](SQL_Answers/Question3Ans.png "Question 3")

## When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

![](SQL_Answers/Question4Script.png "Question 4")
![](SQL_Answers/Question4Ans.png "Question 4")

## Which brand has the most spend among users who were created within the past 6 months?

![](SQL_Answers/Question5Script.png "Question 5")
![](SQL_Answers/Question5Ans.png "Question 5")

## Which brand has the most transactions among users who were created within the past 6 months?

![](SQL_Answers/Question6Script.png "Question 6")
![](SQL_Answers/Question6Ans.png "Question 6")

# Third: Evaluate Data Quality Issues in the Data Provided

Receipt table has a lot of duplicated values as we exploded the dataframe to manage a list item and converting it to a row item with replicating other col values

In [48]:
receipts_final.duplicated(subset='_id').sum()

6262

#### We need more information for handling this type of data as, for the 1st item on the list, it says ITEM NOT FOUND and we have a valid item for second item on the list.


In [49]:
receipts_df['rewardsReceiptItemList'][1]

[{'barcode': '4011',
  'description': 'ITEM NOT FOUND',
  'finalPrice': '1',
  'itemPrice': '1',
  'partnerItemId': '1',
  'quantityPurchased': 1},
 {'barcode': '028400642255',
  'description': 'DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCED FAT BAG 1 OZ',
  'finalPrice': '10.00',
  'itemPrice': '10.00',
  'needsFetchReview': True,
  'needsFetchReviewReason': 'USER_FLAGGED',
  'partnerItemId': '2',
  'pointsNotAwardedReason': 'Action not allowed for user and CPG',
  'pointsPayerId': '5332f5fbe4b03c9a25efd0ba',
  'preventTargetGapPoints': True,
  'quantityPurchased': 1,
  'rewardsGroup': 'DORITOS SPICY SWEET CHILI SINGLE SERVE',
  'rewardsProductPartnerId': '5332f5fbe4b03c9a25efd0ba',
  'userFlaggedBarcode': '028400642255',
  'userFlaggedDescription': 'DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCED FAT BAG 1 OZ',
  'userFlaggedNewItem': True,
  'userFlaggedPrice': '10.00',
  'userFlaggedQuantity': 1}]

#### Whereas, for some rows we have multiple duplicate entries for the same record. As shown below

In [50]:
receipts_df['rewardsReceiptItemList'][12]

[{'barcode': '034100573065',
  'description': 'MILLER LITE 24 PACK 12OZ CAN',
  'finalPrice': '29',
  'itemPrice': '29',
  'partnerItemId': '1',
  'pointsEarned': '870.0',
  'pointsPayerId': '5332f709e4b03c9a25efd0f1',
  'quantityPurchased': 1,
  'rewardsGroup': 'MILLER LITE 24 PACK',
  'rewardsProductPartnerId': '5332f709e4b03c9a25efd0f1',
  'targetPrice': '77'},
 {'barcode': '034100573065',
  'description': 'MILLER LITE 24 PACK 12OZ CAN',
  'finalPrice': '29',
  'itemPrice': '29',
  'partnerItemId': '2',
  'pointsEarned': '870.0',
  'pointsPayerId': '5332f709e4b03c9a25efd0f1',
  'quantityPurchased': 1,
  'rewardsGroup': 'MILLER LITE 24 PACK',
  'rewardsProductPartnerId': '5332f709e4b03c9a25efd0f1',
  'targetPrice': '77'},
 {'barcode': '034100573065',
  'description': 'MILLER LITE 24 PACK 12OZ CAN',
  'finalPrice': '29',
  'itemPrice': '29',
  'partnerItemId': '3',
  'pointsEarned': '870.0',
  'pointsPayerId': '5332f709e4b03c9a25efd0f1',
  'quantityPurchased': 1,
  'rewardsGroup': 'MI

In [51]:
brands_final.duplicated(subset='_id').sum()

0

##### We cannot allow duplicate value for this column as this will serve as a primary key columns for table Brands

In [52]:
brands_final.duplicated(subset='brandCode').sum()

269

##### We cannot allow duplicate value for this column as this will serve as a primary key columns for table Users

In [53]:
users_final.duplicated(subset='_id').sum()

283

##### Difference between unique brands on receipt and brand dataset has a huge difference, as many brand name are not on the receipt table.

In [54]:
len(set(receipts_final['rewardsReceiptItemList_brandCode'])),len(set(brands_final['brandCode']))

(228, 898)

##### Out of 228 brand with receipt, 186 brands are missing on brand table brandCode.

In [55]:
len(set(receipts_final[~(receipts_final['rewardsReceiptItemList_brandCode'].isin(brands_final['brandCode']))]['rewardsReceiptItemList_brandCode']))

186

##### i.e. Only 42 brands are common in both the table (receipt and brand) 

In [56]:
len(set(brands_final['brandCode']))-len(set(brands_final[~(brands_final['brandCode'].isin(receipts_final['rewardsReceiptItemList_brandCode']))]['brandCode']))

42